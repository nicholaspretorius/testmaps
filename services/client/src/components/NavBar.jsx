import React from "react";
import { Link, withRouter } from "react-router-dom";
import PropTypes from "prop-types";

import auth from "./../services/auth";
import "./NavBar.css";

const titleStyle = {
  fontWeight: "bold"
};

const NavBar = props => {
  const { title, logoutUser, isAuthenticated } = props;

  const signOutAuth0 = () => {
    auth.signOut();
    props.history.replace("/");
  };

  let menu = (
    <div className="navbar-menu">
      <div className="navbar-start">
        <Link to="/about" className="navbar-item" data-testid="nav-about">
          About
        </Link>
      </div>
      <div className="navbar-end">
        <Link to="/register" className="navbar-item" data-testid="nav-register">
          Register
        </Link>
        <Link to="/login" className="navbar-item" data-testid="nav-login">
          Login
        </Link>
        {!auth.isAuthenticated() && (
          // eslint-disable-next-line react/jsx-handler-names
          <span className="navbar-item link" onClick={auth.signIn}>
            Auth0 Sign In
          </span>
        )}
        {auth.isAuthenticated() && (
          <>
            <div className="navbar-item">{auth.getProfile().name}</div>

            <div
              className="navbar-item link"
              onClick={() => {
                signOutAuth0();
              }}
            >
              Auth0 Sign Out
            </div>
          </>
        )}
      </div>
    </div>
  );
  if (isAuthenticated()) {
    menu = (
      <div className="navbar-menu">
        <div className="navbar-start">
          <Link to="/about" className="navbar-item" data-testid="nav-about">
            About
          </Link>
          <Link to="/status" className="navbar-item" data-testid="nav-status">
            User Status
          </Link>
        </div>
        <div className="navbar-end">
          <span onClick={logoutUser} className="navbar-item link" data-testid="nav-logout">
            Logout
          </span>
        </div>
      </div>
    );
  }

  return (
    <nav className="navbar is-dark" role="navigation" aria-label="main navigation">
      <section className="container">
        <div className="navbar-brand">
          <Link to="/" className="navbar-item nav-title" style={titleStyle}>
            {title}
          </Link>
          <span
            className="nav-toggle navbar-burger"
            onClick={() => {
              let toggle = document.querySelector(".nav-toggle");
              let menu = document.querySelector(".navbar-menu");
              toggle.classList.toggle("is-active");
              menu.classList.toggle("is-active");
            }}
          >
            <span />
            <span />
            <span />
          </span>
        </div>
        {menu}
      </section>
    </nav>
  );
};

NavBar.propTypes = {
  title: PropTypes.string.isRequired,
  logoutUser: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.func.isRequired
};

export default withRouter(NavBar);
