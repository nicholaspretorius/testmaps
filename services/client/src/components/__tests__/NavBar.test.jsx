import React from "react";
import { cleanup, wait } from "@testing-library/react";

import NavBar from "./../NavBar";

afterEach(cleanup);

describe("when not authenticated", () => {
  const props = {
    title: "Hello, world!",
    isAuth: jest.fn().mockImplementation(() => false),
    profile: null,
    signIn: () => true,
    signOut: () => false
  };

  it("renders a title", () => {
    const { getByText } = renderWithRouter(<NavBar {...props} />);
    expect(getByText(props.title)).toHaveClass("nav-title");
  });

  it("renders the default props", async () => {
    const { getByText, findByTestId } = renderWithRouter(<NavBar {...props} />);
    expect(getByText(props.title)).toHaveClass("nav-title");
    await wait(() => {
      // TODO: Investigate why it is running twice... render/componentDidMount
      expect(props.isAuth).toHaveBeenCalledTimes(6);
    });
    expect((await findByTestId("nav-about")).innerHTML).toBe("About");
    expect((await findByTestId("nav-signin")).innerHTML).toBe("SignIn");
    // expect((await findByTestId("nav-login")).innerHTML).toBe("Login");
  });

  it("renders", () => {
    const { asFragment } = renderWithRouter(<NavBar {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("when authenticated", () => {
  const props = {
    title: "Hello, world!",
    isAuth: jest.fn().mockImplementation(() => true),
    profile: { name: "test@test.com" },
    signIn: () => false,
    signOut: () => true
  };

  it("renders a title", () => {
    const { getByText } = renderWithRouter(<NavBar {...props} />);
    expect(getByText(props.title)).toHaveClass("nav-title");
  });

  it("renders the default props", async () => {
    const { getByText, findByTestId } = renderWithRouter(<NavBar {...props} />);
    expect(getByText(props.title)).toHaveClass("nav-title");
    await wait(() => {
      // TODO: Investigate why it is running twice... render/componentDidMount
      expect(props.isAuth).toHaveBeenCalledTimes(6);
    });
    expect((await findByTestId("nav-about")).innerHTML).toBe("About");
    expect((await findByTestId("nav-signout")).innerHTML).toBe("SignOut");
    expect((await findByTestId("nav-sanity")).innerHTML).toBe("Sanity");
  });

  it("renders", () => {
    const { asFragment } = renderWithRouter(<NavBar {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
