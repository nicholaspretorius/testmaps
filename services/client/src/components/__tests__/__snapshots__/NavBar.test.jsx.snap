// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when authenticated renders 1`] = `
<DocumentFragment>
  <nav
    aria-label="main navigation"
    class="navbar is-dark"
    role="navigation"
  >
    <section
      class="container"
    >
      <div
        class="navbar-brand"
      >
        <a
          class="navbar-item nav-title"
          href="/"
          style="font-weight: bold;"
        >
          Hello, world!
        </a>
        <span
          class="nav-toggle navbar-burger"
        >
          <span />
          <span />
          <span />
        </span>
      </div>
      <div
        class="navbar-menu"
      >
        <div
          class="navbar-start"
        >
          <a
            class="navbar-item"
            data-testid="nav-about"
            href="/about"
          >
            About
          </a>
          <a
            class="navbar-item"
            data-testid="nav-sanity"
            href="/sanity"
          >
            Sanity
          </a>
        </div>
        <div
          class="navbar-end"
        >
          <div
            class="navbar-item"
          >
            test@test.com
          </div>
          <div
            class="navbar-item link"
            data-testid="nav-signout"
          >
            SignOut
          </div>
        </div>
      </div>
    </section>
  </nav>
</DocumentFragment>
`;

exports[`when not authenticated renders 1`] = `
<DocumentFragment>
  <nav
    aria-label="main navigation"
    class="navbar is-dark"
    role="navigation"
  >
    <section
      class="container"
    >
      <div
        class="navbar-brand"
      >
        <a
          class="navbar-item nav-title"
          href="/"
          style="font-weight: bold;"
        >
          Hello, world!
        </a>
        <span
          class="nav-toggle navbar-burger"
        >
          <span />
          <span />
          <span />
        </span>
      </div>
      <div
        class="navbar-menu"
      >
        <div
          class="navbar-start"
        >
          <a
            class="navbar-item"
            data-testid="nav-about"
            href="/about"
          >
            About
          </a>
        </div>
        <div
          class="navbar-end"
        >
          <span
            class="navbar-item link"
            data-testid="nav-signin"
          >
            SignIn
          </span>
        </div>
      </div>
    </section>
  </nav>
</DocumentFragment>
`;
